Spigot-Server-Patches/0594-Significantly-improve-performance-of-the-end-generat.patch

```diff
@@ -0,0 +1,79 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: SuperCoder7979 <25208576+SuperCoder7979@users.noreply.github.com>
+Date: Tue, 3 Nov 2020 23:48:05 -0600
+Subject: [PATCH] Significantly improve performance of the end generation
+
+This patch implements a noise cache for the end which significantly reduces the computation time of generation. This results in about a 3x improvement.
+
+Original code by SuperCoder7979 and Gegy in Lithium, licensed under LGPL-3.0 <Source: https://github.com/jellysquid3/lithium-fabric>
+
+Co-authored-by: Gegy <gegy1000@gmail.com>
+Co-authored-by: Dylan Xaldin <Puremin0rez515@gmail.com>
+Co-authored-by: pop4959 <pop4959@gmail.com>
+
+diff --git a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+index 9c15cd303ccc2928be4e3e7080e74e8314d7e0f2..195a9b53062f7d3896d17210989d6785ea4baee3 100644
+--- a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
++++ b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+@@ -3,6 +3,9 @@ package net.minecraft.server;
+ import com.google.common.collect.ImmutableList;
+ import com.mojang.serialization.Codec;
+ import com.mojang.serialization.codecs.RecordCodecBuilder;
++import it.unimi.dsi.fastutil.HashCommon; // Paper
++import it.unimi.dsi.fastutil.objects.Object2ObjectLinkedOpenHashMap; // Paper
++import it.unimi.dsi.fastutil.objects.Object2ObjectMap; // Paper
+ import java.util.List;
+ 
+ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+@@ -22,6 +25,13 @@ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+     private final BiomeBase k;
+     private final BiomeBase l;
+     private final BiomeBase m;
++    // Paper start
++    private static final class NoiseCache {
++        public Object2ObjectMap<NoiseGenerator3Handler, long[]> keys = new Object2ObjectLinkedOpenHashMap<>();
++        public Object2ObjectMap<NoiseGenerator3Handler, float[]> values = new Object2ObjectLinkedOpenHashMap<>();
++    }
++    private static final ThreadLocal<NoiseCache> noiseCache = ThreadLocal.withInitial(NoiseCache::new);
++    // Paper end
+ 
+     public WorldChunkManagerTheEnd(IRegistry<BiomeBase> iregistry, long i) {
+         this(iregistry, i, (BiomeBase) iregistry.d(Biomes.THE_END), (BiomeBase) iregistry.d(Biomes.END_HIGHLANDS), (BiomeBase) iregistry.d(Biomes.END_MIDLANDS), (BiomeBase) iregistry.d(Biomes.SMALL_END_ISLANDS), (BiomeBase) iregistry.d(Biomes.END_BARRENS));
+@@ -76,13 +86,35 @@ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+ 
+         f = MathHelper.a(f, -100.0F, 80.0F);
+ 
++        // Paper start
++        NoiseCache cache = noiseCache.get();
++        long[] keys = cache.keys.computeIfAbsent(noisegenerator3handler, m -> {
```

Comment:
Might be beneficial to avoid the double map-lookup by storing a separate object that contains the keys and values?